dft/src/radix_2_dit_parallel.rs

```diff
@@ -21,7 +21,10 @@ use crate::{reverse_bits, reverse_slice_index_bits, TwoAdicSubgroupDft};
 pub struct Radix2DitParallel;
 
 impl<F: TwoAdicField> TwoAdicSubgroupDft<F> for Radix2DitParallel {
-    fn dft_batch(&self, mut mat: RowMajorMatrix<F>) -> RowMajorMatrix<F> {
+    fn dft_batch_bitrev<const IN_BITREV: bool, const OUT_BITREV: bool>(
```

Comment:
The idea is that Dft implementers can just override the `bitrev` methods, and statically choose whether to bit-reverse inputs and outputs when necessary. It doesn't seem to save much now, but may help squeeze out milliseconds later.

Reply:
having it statically dispatched isn't *that* important, though, since it's just a conditional jump.. I just thought the API was nicer. concern would be if it pressures icache more if switching between versions..

Reply:
Looks reasonable!

I think ordering probably affects LDE costs more than other prover costs, so there might be some advantage to letting the LDE impl decide the output ordering and have FRI logic lazily reindex if needed. Though I realize there might be complications with that in Rust.

cc @SyxtonPrime for visibility since he's working on a (circle group related) FFT impl, though I'm not sure if it would share the same traits anyway.

Reply:
Ahh hm I see. I was focused on keeping the important path (Radix2DitParallel + FRI) fast, but the important path can change. I can brainstorm some options for a matrix reindexing API.

Reply:
I also didn't know if it was important to keep Dft taking actual `RowMajorMatrix` instead of a trait like `MatrixRowSlices`, in case of optimizations down the line. Haven't ever written an optimized fft myself.

Reply:
well, I guess it's mostly to reuse the Vec.

Reply:
I guess if the magical `ReindexedMatrix` can be `MatrixRowSlicesMut`, the Dft can still just write directly to the reindexed slices, and optionally "reindex" again at the end (possibly reverting to the original identity index). Splitting the rows for rayon *might* be tricky. I feel like it would either be very clean and nice or horribly complex, no in between

Reply:
Looks like llvm doesn't optimize `rbit` super well, so we can't rely on it to get rid of reindexing overhead: https://godbolt.org/z/q1jWTWjYx

Must-haves:
1. Index-transforming function is statically determined so matrix accesses aren't glacial
2. `Dft` or others can call `.to_row_major_matrix()` to realize a dense repr in "true" order, possibly copying
3. Realizing the dense repr of $2k$ bit reverses is a no-op

Nice-to-haves, in descending order:
4. Does not pollute trait bounds everywhere
5. Can handle arbitrary row index transform so as to absorb VerticallyStrided and possibly VerticalStack
6. Can handle arbitrary row, col transform to provide a lazy Transpose API

The problem with 5 & 6 is their interaction with 1

Very rough sketch:
```rust
trait RowOrdering {
    fn transform_index(idx: usize) -> usize;
}

struct BitRev;
impl RowOrdering for BitRev { ... }

struct MatrixView<T, Order>(/* Cow or something, tbd */)
impl<T, Order: RowOrdering> MatrixView<T, Order> {
    fn reorder<Order2>(self) -> MatrixView<T, ???> {
        // if bitrev -> bitrev, IdentityOrder
        // otherwise Compose<Order2, Order>
    }
}
```

The only option I see is to have $n^2$ structs for `CanReorder::<Order1, Order2>::OutputOrder`, which is not only annoying to define but also annoying to propagate through all users' type bounds.

So uhhhh I will keep thinking of a better way

Reply:
so yeah I will pass on designing the uber-API rn and just set it up for lazy bit-reversal

Reply:
I suppose it's okay for the index-mapping to be done at runtime (un-optimizable at compile time) if it's only for row reordering. It would slow down views of narrow matrices slightly but be negligible for wider matrices, assuming most work is done on the whole row